### PR TITLE
Fix partial cloning a repo

### PR DESCRIPTION
**Pull request by Gusted**
- Pass the Global command args into serviceRPC.
- Fixes error while partial cloning.
- Added test to ensure partial cloning functionality.

### CODE REVIEW

**Review by lunny — APPROVED** [review]

**Review by KN4CK3R — APPROVED** [review]

**Comment by lunny** [comment]
But it seems before #18363, global command args are also not added?

**Comment by Gusted** [comment]
> But it seems before #18363, global command args are also not added?

Yes, that's indeed the case... git bisect does point out to that particular comment, so something must have changed, most likely `git.RunWithContext` is adding something to git to make it more strict of some sort. Really weird regression tho.

**Comment by lunny** [comment]
Wait, I will add some tests for this PR.

**Comment by zeripath** [comment]
On the whole I'm tempted to say we should revert #18363 and represent it once we know what's really going on here. 

I'm also suspicious that the control flow in services/gitdiff/gitdiff.go reader isn't right.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@zeripath I think we can merge this and look into control flow of reader in a follow up ?

File: routers/web/repo/http.go [hunk 1]
```diff
@@ -485,7 +485,7 @@ func serviceRPC(ctx gocontext.Context, h serviceHandler, service string) {
 	}
 
 	var stderr bytes.Buffer
-	cmd := git.NewCommandContextNoGlobals(h.r.Context(), service, "--stateless-rpc", h.dir)
```
Comment (6543) [inline routers/web/repo/http.go:488]: this looks valid before - blocking to have a deeper dig into code ... cc @zeripath

**Review by 6543 — DISMISSED** [review]

File: routers/web/repo/http.go (hunk 1 above)
Comment (zeripath) [reply routers/web/repo/http.go:488]: I just don't understand why this ever worked and why this fixes things.

**Comment by Gusted** [comment]
Well to add more confusion to this niece mystery, now the `--filter=` command is even broken before the PR of lunny, I'm not sure what I did, but it might not be a regression after all? Could someone confirm this on a local instance by doing this on a recent main branch and on https://github.com/go-gitea/gitea/commit/f066b293accb8aed5809de87ae56140c90978d66.

Command:
```
git clone --filter=blob:none http://localhost:3000/owner/repo
```

File: routers/web/repo/http.go (hunk 1 above)
Comment (Gusted) [reply routers/web/repo/http.go:488]: > I just don't understand why this ever worked

See my recent comment.



> and why this fixes things.

The `NewCommandContext` function add the globals, one of them is to allow the filter functionality.

https://github.com/go-gitea/gitea/blob/5e5740af69cb0f44ff0c1576a2387388d75bb4c2/modules/git/git.go#L149-lL152

Comment (6543) [reply routers/web/repo/http.go:488]: so it's not a regression - its a independent fix

**Comment by zeripath** [comment]
OK I think I  can now understand why your thing fixes this.

Adding the Globals means that:  "-c", "uploadpack.allowfilter=true" is added.

File: routers/web/repo/http.go (hunk 1 above)
Comment (Gusted) [reply routers/web/repo/http.go:488]: > so it's not a regression - its a independent fix

Seems like it, I've changed the Title + Comment according to it.

**Comment by zeripath** [comment]
Now the issue with the Diff is that I think we also need to be setting the Globals there too.

**Review by 6543 — APPROVED** [review]
ACK as addressed correctly and confusion solved

**Comment by Gusted** [comment]
> Now the issue with the Diff is that I think we also need to be setting the Globals there too.

Hmm, which of the global arguments would change the behavior of diff? I don't think those are impacted by those arguments, but we could include them in order to be future-proof.

**Comment by zeripath** [comment]
The lfs.filters I'm afraid.

**Review by zeripath — APPROVED** [review]

**Pull request merged**